File: fingerprint_backend/app/services/template_encoder.py
import struct
import base64
import numpy as np
from dataclasses import dataclass
from typing import Optional, List

# ── Constants ─────────────────────────────────────────────────────────────────
_ENDING          = 1
_BIFURCATION     = 2
_BORDER          = 15
_CLUSTER_R       = 12.0
_MAX_MINUTIAE    = 80
_MIN_MINUTIAE    = 10
_HEADER_SIZE     = 30    # bytes before minutiae array
_MIN_RIDGE_PX    = 12    # min skeleton pixels in neighbourhood — filters noise spikes
_FOLLOW_STEPS    = 7     # pixels to follow per branch for angle estimation
_GRID_ROWS       = 4     # spatial distribution grid
_GRID_COLS       = 4
_MAX_PER_CELL    = 5     # max minutiae per grid cell


@dataclass
class Minutia:
    x:       int
    y:       int
    angle:   int   # 0-255  (maps to 0-360°)
    type:    int   # 1=ending  2=bifurcation
    quality: int   # 0-100


# ── Public API ────────────────────────────────────────────────────────────────

def encode(
    skeleton: np.ndarray,
    finger_position: int = 0,
    quality_score: float = 80.0
) -> Optional[str]:
    """
    Extract minutiae from a skeletonised image and return a
    base64-encoded ISO 19794-2 Finger Minutiae Record.
    Returns None when ridge detail is insufficient (< MIN_MINUTIAE).
    """
    minutiae = _extract(skeleton)
    filtered = _filter(minutiae, skeleton.shape[1], skeleton.shape[0])

    if len(filtered) < _MIN_MINUTIAE:
        return None

    return _build_template(
        filtered,
        skeleton.shape[1], skeleton.shape[0],
        finger_position,
        int(np.clip(quality_score, 0, 100))
    )


def parse(b64_template: str) -> List[Minutia]:
    """
    Parse a base64-encoded ISO 19794-2 template back to a list of Minutia.
    Used by the matcher.
    """
    data = base64.b64decode(b64_template)
    if len(data) < _HEADER_SIZE:
        return []

    count = data[29]          # minutiae count is at byte 29
    minutiae: List[Minutia] = []
    offset = _HEADER_SIZE

    for _ in range(count):
        if offset + 6 > len(data):
            break
        b0, b1, b2, b3, angle, quality = data[offset:offset + 6]

        mtype = (b0 >> 6) & 0x03
        x     = ((b0 & 0x3F) << 8) | b1
        y     = ((b2 & 0x3F) << 8) | b3

        minutiae.append(Minutia(x, y, angle, mtype, quality))
        offset += 6

    return minutiae


# ── Minutiae extraction ───────────────────────────────────────────────────────

def _extract(skeleton: np.ndarray) -> List[Minutia]:
    rows, cols = skeleton.shape
    minutiae: List[Minutia] = []

    def px(r: int, c: int) -> int:
        if r < 0 or r >= rows or c < 0 or c >= cols:
            return 0
        return 1 if skeleton[r, c] > 127 else 0

    for r in range(_BORDER, rows - _BORDER):
        for c in range(_BORDER, cols - _BORDER):
            if px(r, c) == 0:
                continue

            n = [
                px(r-1, c-1), px(r-1, c), px(r-1, c+1),
                px(r,   c+1),
                px(r+1, c+1), px(r+1, c), px(r+1, c-1),
                px(r,   c-1)
            ]
            cn = sum(abs(n[i] - n[(i + 1) % 8]) for i in range(8)) // 2

            if cn == 1:
                mtype = _ENDING
            elif cn == 3:
                mtype = _BIFURCATION
            else:
                continue

            # Reject minutiae on very short ridge stubs (noise artifacts)
            if not _has_sufficient_ridge(skeleton, r, c, rows, cols):
                continue

            angle   = _ridge_angle(skeleton, r, c, rows, cols)
            quality = _local_quality(skeleton, r, c, rows, cols)
            minutiae.append(Minutia(c, r, angle, mtype, quality))

    return minutiae


def _follow_ridge(skeleton: np.ndarray, start_r: int, start_c: int,
                  from_r: int, from_c: int, rows: int, cols: int) -> tuple:
    """
    Walk the skeleton ridge from (start_r, start_c) away from (from_r, from_c)
    for up to _FOLLOW_STEPS pixels.  Returns the endpoint (end_r, end_c).

    Used by _ridge_angle to get a stable direction vector instead of a noisy
    single-pixel gradient.
    """
    def px(rr: int, cc: int) -> bool:
        return 0 <= rr < rows and 0 <= cc < cols and skeleton[rr, cc] > 127

    prev_r, prev_c = from_r, from_c
    cur_r,  cur_c  = start_r, start_c

    for _ in range(_FOLLOW_STEPS):
        moved = False
        # 4-connected first (prefer straight path), then diagonal
        for dr, dc in [(-1,0),(1,0),(0,-1),(0,1),(-1,-1),(-1,1),(1,-1),(1,1)]:
            nr, nc = cur_r + dr, cur_c + dc
            if (nr, nc) != (prev_r, prev_c) and px(nr, nc):
                prev_r, prev_c = cur_r, cur_c
                cur_r, cur_c   = nr, nc
                moved = True
                break
        if not moved:
            break

    return cur_r, cur_c


def _ridge_angle(skeleton: np.ndarray, r: int, c: int, rows: int, cols: int) -> int:
    """
    Estimate minutia angle by following each ridge branch for _FOLLOW_STEPS pixels
    and averaging the resulting direction vectors.

    WHY THIS IS BETTER than a single-point gradient:
    A single gradient pixel is highly sensitive to skeleton noise — one stray
    pixel can flip the angle by 90°+.  Following the ridge several steps smooths
    out local perturbations and gives a direction that reflects the true ridge
    trajectory, which is what ISO 19794-2 angle represents.
    """
    def px(rr: int, cc: int) -> bool:
        return 0 <= rr < rows and 0 <= cc < cols and skeleton[rr, cc] > 127

    neighbors = [
        (r + dr, c + dc)
        for dr in (-1, 0, 1) for dc in (-1, 0, 1)
        if not (dr == 0 and dc == 0) and px(r + dr, c + dc)
    ]

    if not neighbors:
        return 0

    dx_sum, dy_sum = 0.0, 0.0
    for nr, nc in neighbors:
        end_r, end_c = _follow_ridge(skeleton, nr, nc, r, c, rows, cols)
        dx_sum += end_c - c
        dy_sum += end_r - r

    angle_deg = float(np.degrees(np.arctan2(dy_sum, dx_sum)))
    return int(((angle_deg + 360.0) % 360.0) / 360.0 * 255.0)


def _local_quality(skeleton: np.ndarray, r: int, c: int, rows: int, cols: int) -> int:
    """
    Local ridge quality score (0-100).

    Measures skeleton pixel density in a 24×24 neighbourhood.
    A minutia on a strong, continuous ridge has high density; one on a short
    noise stub has few surrounding pixels → low score.

    This replaces the Laplacian magnitude which measured edge sharpness
    (always high on the skeleton by construction) rather than actual ridge quality.
    """
    r1 = max(0, r - 12);  r2 = min(rows, r + 12)
    c1 = max(0, c - 12);  c2 = min(cols, c + 12)
    area    = max(1, (r2 - r1) * (c2 - c1))
    nonzero = int(np.count_nonzero(skeleton[r1:r2, c1:c2] > 127))
    # Dense ridge area: nonzero/area ≈ 0.4-0.6 for a good skeleton crop
    return int(np.clip(nonzero / area * 200.0, 0, 100))


def _has_sufficient_ridge(skeleton: np.ndarray, r: int, c: int,
                          rows: int, cols: int) -> bool:
    """
    Return True only if there are at least _MIN_RIDGE_PX skeleton pixels in the
    immediate neighbourhood of (r, c).

    WHY:  The crossing-number algorithm fires on single isolated pixels and
    tiny noise stubs that survived skeletonization.  These produce false minutiae
    that degrade matching.  A real ridge ending or bifurcation always sits on a
    ridge with many surrounding pixels.
    """
    r1 = max(0, r - 8);  r2 = min(rows, r + 8)
    c1 = max(0, c - 8);  c2 = min(cols, c + 8)
    return int(np.count_nonzero(skeleton[r1:r2, c1:c2] > 127)) >= _MIN_RIDGE_PX


# ── Minutiae filtering ────────────────────────────────────────────────────────

def _filter(raw: List[Minutia], width: int, height: int) -> List[Minutia]:
    """
    Filter raw minutiae list to a clean, well-distributed set.

    Three passes in priority order (all operating on quality-sorted input):
      1. Border exclusion  — discard minutiae too close to crop edges
      2. Proximity cluster — keep only one minutia per _CLUSTER_R radius
      3. Spatial grid cap  — limit to _MAX_PER_CELL per grid cell so minutiae
                             are spread across the fingerprint rather than
                             clumped in one high-texture region

    WHY SPATIAL DISTRIBUTION:
    If 60 of 80 minutiae fall in the upper-left quarter, the matcher only has
    meaningful overlap there.  Enforcing a grid cap ensures the template
    represents the whole fingerprint, which dramatically improves match rate
    when the user presents the finger at a slightly different position.
    """
    sorted_m = sorted(raw, key=lambda m: m.quality, reverse=True)
    accepted:    List[Minutia] = []
    cell_counts: dict          = {}

    cell_w = max(1, width  / _GRID_COLS)
    cell_h = max(1, height / _GRID_ROWS)

    for m in sorted_m:
        # 1. Border exclusion
        if not (_BORDER <= m.x <= width  - _BORDER):
            continue
        if not (_BORDER <= m.y <= height - _BORDER):
            continue

        # 2. Proximity cluster
        if any(np.hypot(m.x - e.x, m.y - e.y) < _CLUSTER_R for e in accepted):
            continue

        # 3. Spatial grid cap
        cell = (int(m.y / cell_h), int(m.x / cell_w))
        if cell_counts.get(cell, 0) >= _MAX_PER_CELL:
            continue

        cell_counts[cell] = cell_counts.get(cell, 0) + 1
        accepted.append(m)

        if len(accepted) >= _MAX_MINUTIAE:
            break

    return accepted


# ── ISO 19794-2 binary builder ────────────────────────────────────────────────

def _build_template(
    minutiae: List[Minutia],
    width: int, height: int,
    finger_position: int,
    finger_quality: int
) -> str:
    count = min(len(minutiae), _MAX_MINUTIAE)
    record_length = _HEADER_SIZE + count * 6 + 2

    buf = bytearray(record_length)

    # Format identifier + version
    buf[0:4]  = b'FMR\x00'
    buf[4:8]  = b' 20\x00'

    # Record length, CBEFF, equipment, dimensions, resolution
    struct.pack_into('>I', buf,  8, record_length)
    struct.pack_into('>H', buf, 12, 0)        # CBEFF product ID
    struct.pack_into('>H', buf, 14, 0)        # capture equipment
    struct.pack_into('>H', buf, 16, width)
    struct.pack_into('>H', buf, 18, height)
    struct.pack_into('>H', buf, 20, 197)      # X resolution (~500 dpi)
    struct.pack_into('>H', buf, 22, 197)      # Y resolution

    buf[24] = 1                               # number of views
    buf[25] = 0                               # reserved
    buf[26] = finger_position & 0xFF
    buf[27] = 0x00                            # view 0, live-scan plain
    buf[28] = finger_quality & 0xFF
    buf[29] = count & 0xFF

    offset = _HEADER_SIZE
    for m in minutiae[:count]:
        type_x_high = ((m.type & 0x03) << 6) | ((m.x >> 8) & 0x3F)
        buf[offset]     = type_x_high & 0xFF
        buf[offset + 1] = m.x & 0xFF
        buf[offset + 2] = (m.y >> 8) & 0x3F
        buf[offset + 3] = m.y & 0xFF
        buf[offset + 4] = m.angle & 0xFF
        buf[offset + 5] = m.quality & 0xFF
        offset += 6

    # No extended data
    struct.pack_into('>H', buf, offset, 0)

    return base64.b64encode(bytes(buf)).decode('utf-8')

File: fingerprint-sdk/fingerprint_sdk/android/src/main/cpp/template_encoder.cpp
#include "template_encoder.h"
#include "fingerprint_core.h"
#include "quality_analyzer.h"
#include "liveness_detector.h"
#include <cmath>
#include <algorithm>
#include <map>

// ---------------------------------------------------------------------------
// Base64 helpers (self-contained, no external dependency)
// ---------------------------------------------------------------------------
static const std::string B64_CHARS =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

static std::string b64Encode(const std::vector<uint8_t>& data) {
    std::string out;
    out.reserve(((data.size() + 2) / 3) * 4);
    int val = 0, valb = -6;
    for (uint8_t c : data) {
        val = (val << 8) + c;
        valb += 8;
        while (valb >= 0) {
            out.push_back(B64_CHARS[(val >> valb) & 0x3F]);
            valb -= 6;
        }
    }
    if (valb > -6)
        out.push_back(B64_CHARS[((val << 8) >> (valb + 8)) & 0x3F]);
    while (out.size() % 4) out.push_back('=');
    return out;
}

static std::vector<uint8_t> b64Decode(const std::string& s) {
    std::vector<uint8_t> out;
    std::vector<int> T(256, -1);
    for (int i = 0; i < 64; i++) T[(unsigned char)B64_CHARS[i]] = i;
    int val = 0, valb = -8;
    for (unsigned char c : s) {
        if (T[c] == -1) break;
        val = (val << 6) + T[c];
        valb += 6;
        if (valb >= 0) {
            out.push_back((val >> valb) & 0xFF);
            valb -= 8;
        }
    }
    return out;
}

// ---------------------------------------------------------------------------

namespace fingerprint {

int TemplateEncoder::crossingNumber(const cv::Mat& skel, int x, int y) {
    // 8-neighbour crossing number; CN=1 → ending, CN=3 → bifurcation
    uint8_t nb[9];
    nb[1] = skel.at<uint8_t>(y-1, x  ) > 0 ? 1 : 0;
    nb[2] = skel.at<uint8_t>(y-1, x+1) > 0 ? 1 : 0;
    nb[3] = skel.at<uint8_t>(y,   x+1) > 0 ? 1 : 0;
    nb[4] = skel.at<uint8_t>(y+1, x+1) > 0 ? 1 : 0;
    nb[5] = skel.at<uint8_t>(y+1, x  ) > 0 ? 1 : 0;
    nb[6] = skel.at<uint8_t>(y+1, x-1) > 0 ? 1 : 0;
    nb[7] = skel.at<uint8_t>(y,   x-1) > 0 ? 1 : 0;
    nb[8] = skel.at<uint8_t>(y-1, x-1) > 0 ? 1 : 0;

    int cn = 0;
    for (int k = 1; k <= 8; k++)
        cn += std::abs((int)nb[k] - (int)nb[k % 8 + 1]);
    return cn / 2;
}

float TemplateEncoder::followRidge(const cv::Mat& skel,
                                   int sx, int sy,
                                   int px, int py,
                                   int steps) {
    static const int dx8[8] = {0, 1, 1, 1, 0,-1,-1,-1};
    static const int dy8[8] = {-1,-1,0, 1, 1, 1, 0,-1};

    int x = sx, y = sy, ppx = px, ppy = py;
    float totalAngle = 0.0f;

    for (int s = 0; s < steps; s++) {
        int nx = -1, ny = -1;
        for (int d = 0; d < 8; d++) {
            int cx = x + dx8[d];
            int cy = y + dy8[d];
            if (cx == ppx && cy == ppy) continue;
            if (cx < 0 || cy < 0 || cx >= skel.cols || cy >= skel.rows) continue;
            if (skel.at<uint8_t>(cy, cx) > 0) { nx = cx; ny = cy; break; }
        }
        if (nx == -1) break;
        totalAngle += std::atan2((float)(ny - y), (float)(nx - x));
        ppx = x; ppy = y;
        x = nx; y = ny;
    }
    return totalAngle;
}

std::vector<Minutia>
TemplateEncoder::extractMinutiae(const cv::Mat& skel) {
    std::vector<Minutia> result;
    const int margin = 15;

    for (int y = margin; y < skel.rows - margin; y++) {
        for (int x = margin; x < skel.cols - margin; x++) {
            if (skel.at<uint8_t>(y, x) == 0) continue;

            int cn = crossingNumber(skel, x, y);
            if (cn != 1 && cn != 3) continue;

            // Ridge support: need >=12 skeleton pixels in 16×16 neighbourhood
            int x0 = std::max(0, x - 8), y0 = std::max(0, y - 8);
            cv::Rect roi16(x0, y0,
                           std::min(16, skel.cols - x0),
                           std::min(16, skel.rows - y0));
            if (cv::countNonZero(skel(roi16)) < 12) continue;

            // Follow ridge 7 steps to get stable angle
            float rawAngle = followRidge(skel, x, y, x - 1, y, 7);
            float angle = (float)std::fmod(rawAngle + 2.0 * CV_PI, 2.0 * CV_PI);

            // Local density in 24×24 window → quality
            int qx = std::max(0, x - 12), qy = std::max(0, y - 12);
            cv::Rect roi24(qx, qy,
                           std::min(24, skel.cols - qx),
                           std::min(24, skel.rows - qy));
            int density = cv::countNonZero(skel(roi24));
            uint8_t quality = (uint8_t)std::min(100, density * 2);

            Minutia m;
            m.x       = x;
            m.y       = y;
            m.angle   = angle;
            m.quality = quality;
            m.type    = (cn == 1) ? MinutiaeType::ENDING : MinutiaeType::BIFURCATION;
            result.push_back(m);
        }
    }
    return result;
}

std::vector<Minutia>
TemplateEncoder::filterMinutiae(const std::vector<Minutia>& raw,
                                int width, int height) {
    // 1. Non-maximum suppression within 12 px radius
    std::vector<bool> suppressed(raw.size(), false);
    for (size_t i = 0; i < raw.size(); i++) {
        if (suppressed[i]) continue;
        for (size_t j = i + 1; j < raw.size(); j++) {
            if (suppressed[j]) continue;
            float dx = (float)(raw[i].x - raw[j].x);
            float dy = (float)(raw[i].y - raw[j].y);
            if (dx * dx + dy * dy < 144.0f) {
                size_t worst = (raw[i].quality < raw[j].quality) ? i : j;
                suppressed[worst] = true;
            }
        }
    }

    std::vector<Minutia> filtered;
    for (size_t i = 0; i < raw.size(); i++)
        if (!suppressed[i]) filtered.push_back(raw[i]);

    // 2. Spatial distribution: max 5 per grid cell (4×4 grid)
    int cellW = std::max(1, width  / 4);
    int cellH = std::max(1, height / 4);

    std::sort(filtered.begin(), filtered.end(),
              [](const Minutia& a, const Minutia& b){
                  return a.quality > b.quality;
              });

    std::map<std::pair<int,int>, int> cellCount;
    std::vector<Minutia> spatial;
    for (const auto& m : filtered) {
        auto cell = std::make_pair(m.x / cellW, m.y / cellH);
        if (cellCount[cell] < 5) {
            spatial.push_back(m);
            cellCount[cell]++;
        }
    }

    // 3. Hard cap at 80 minutiae (ISO limit)
    if (spatial.size() > 80) spatial.resize(80);
    return spatial;
}

std::string
TemplateEncoder::encodeISO19794(const std::vector<Minutia>& minutiae,
                                int width, int height) {
    // ISO 19794-2:2005  — simplified FMR binary layout
    // 30-byte header + 6 bytes per minutia
    int totalBytes = 30 + 6 * (int)minutiae.size();
    std::vector<uint8_t> buf(totalBytes, 0);

    // Magic "FMR\0" + version " 20\0"
    buf[0]='F'; buf[1]='M'; buf[2]='R'; buf[3]=0;
    buf[4]=' '; buf[5]='2'; buf[6]='0'; buf[7]=0;

    // Total record length (4 bytes big-endian)
    buf[8]  = (totalBytes >> 24) & 0xFF;
    buf[9]  = (totalBytes >> 16) & 0xFF;
    buf[10] = (totalBytes >>  8) & 0xFF;
    buf[11] =  totalBytes        & 0xFF;

    // Reserved 2 bytes
    buf[12] = buf[13] = 0;

    // Image dimensions
    buf[14] = (width  >> 8) & 0xFF;  buf[15] = width  & 0xFF;
    buf[16] = (height >> 8) & 0xFF;  buf[17] = height & 0xFF;

    // Resolution: 197 ppi
    buf[18] = 0; buf[19] = 197;
    buf[20] = 0; buf[21] = 197;

    buf[22] = 1;    // number of finger views
    buf[23] = 0;    // view number
    buf[24] = 0;    // impression type (live plain)
    buf[25] = 80;   // finger quality
    buf[26] = (uint8_t)minutiae.size();
    buf[27] = buf[28] = buf[29] = 0; // reserved

    // Minutiae records (6 bytes each)
    int offset = 30;
    for (const auto& m : minutiae) {
        uint8_t typeCode = (m.type == MinutiaeType::ENDING) ? 1 : 3;

        // Bytes 0-1: type[2] | x[14]
        uint16_t xt = ((uint16_t)typeCode << 14) | (uint16_t)(m.x & 0x3FFF);
        buf[offset]   = (xt >> 8) & 0xFF;
        buf[offset+1] = xt        & 0xFF;

        // Bytes 2-3: y[14] (2 reserved bits = 0)
        uint16_t yt = (uint16_t)(m.y & 0x3FFF);
        buf[offset+2] = (yt >> 8) & 0xFF;
        buf[offset+3] = yt        & 0xFF;

        // Byte 4: angle mapped 0-255
        buf[offset+4] = (uint8_t)(m.angle / (2.0f * (float)CV_PI) * 255.0f);

        // Byte 5: quality
        buf[offset+5] = m.quality;
        offset += 6;
    }

    return b64Encode(buf);
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

TemplateResult TemplateEncoder::extractTemplate(const cv::Mat& image,
                                                int w, int h) {
    TemplateResult result{};
    result.success = false;

    int width  = (w > 0) ? w : image.cols;
    int height = (h > 0) ? h : image.rows;

    cv::Mat skeleton = ImageProcessor::enhanceForEncoding(image);
    if (skeleton.empty()) {
        result.errorMessage = "Skeletonization failed";
        return result;
    }

    auto rawMinutiae = extractMinutiae(skeleton);
    if ((int)rawMinutiae.size() < 10) {
        result.errorMessage = "Insufficient minutiae: " +
                               std::to_string(rawMinutiae.size());
        return result;
    }

    result.minutiae = filterMinutiae(rawMinutiae, width, height);
    if ((int)result.minutiae.size() < 10) {
        result.errorMessage = "Insufficient minutiae after filtering: " +
                               std::to_string(result.minutiae.size());
        return result;
    }

    result.base64Template = encodeISO19794(result.minutiae, width, height);
    result.minutiaeCount  = (int)result.minutiae.size();
    result.success        = true;
    return result;
}

std::vector<Minutia>
TemplateEncoder::decodeTemplate(const std::string& base64Template) {
    std::vector<Minutia> out;
    auto buf = b64Decode(base64Template);
    if ((int)buf.size() < 30) return out;

    int numMinutiae   = buf[26];
    int expectedBytes = 30 + 6 * numMinutiae;
    if ((int)buf.size() < expectedBytes) return out;

    for (int i = 0; i < numMinutiae; i++) {
        int offset = 30 + i * 6;
        Minutia m;

        uint16_t xt      = ((uint16_t)buf[offset] << 8) | buf[offset+1];
        uint8_t  typeCode = (uint8_t)((xt >> 14) & 0x3);
        m.x    = xt & 0x3FFF;
        m.type = (typeCode == 1) ? MinutiaeType::ENDING : MinutiaeType::BIFURCATION;

        uint16_t yt = ((uint16_t)buf[offset+2] << 8) | buf[offset+3];
        m.y = yt & 0x3FFF;

        m.angle   = buf[offset+4] / 255.0f * 2.0f * (float)CV_PI;
        m.quality = buf[offset+5];

        out.push_back(m);
    }
    return out;
}

} // namespace fingerprint
